libfoundation/src/foundation-unicode.cpp

```diff
@@ -771,80 +771,142 @@ bool MCUnicodeIsNormalisedNKD(const unichar_t *p_string, uindex_t p_length)
 
 ////////////////////////////////////////////////////////////////////////////////
 
-bool MCUnicodeLowercase(MCLocaleRef p_locale,
+static bool __MCUnicodeMapString(MCUnicodeProperty p_property, MCLocaleRef p_locale,
                         const unichar_t *p_in, uindex_t p_in_length,
                         unichar_t *&r_out, uindex_t &r_out_length)
 {
+    MCAutoArray<unichar_t> t_buffer;
+    uindex_t t_out_length;
+    
     // Do the casing using icu::UnicodeString. Note that a locale is required.
     UErrorCode t_error = U_ZERO_ERROR;
     icu::UnicodeString t_input(p_in, p_in_length);
-    t_input.toLower(MCLocaleGetICULocale(p_locale));
+    
+    switch (p_property)
+    {
+        case kMCUnicodePropertyLowercaseMapping:
+        case kMCUnicodePropertyUppercaseMapping:
+        case kMCUnicodePropertyTitlecaseMapping:
+        case kMCUnicodePropertyCaseFolding:
+        {
+            icu::Locale t_locale = nullptr;
+            if (p_locale != nullptr)
+            {
+                t_locale = MCLocaleGetICULocale(p_locale);
```

Comment:
@runrevmark maybe you deleted your comment but I got the email anyway? Anyway no this does not leak it's just an abstraction to access the icu locale in the `__MCLocale` struct.

Reply:
Indeed - that's why I deleted the comment - I missed the fact the locale-ref was passed in first time around :)